### PR TITLE
Write oldest documents to Index first

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This PR:

- Simplifies the index writing logic by capping at the number of documents instead of index entries written. This also avoids re-reading the same document multiple times.
- Adds logic to write the documents in order of read time.

### CODE REVIEW

**Comment by schmidt-sebastian** [comment]
cc @wu-hui

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/IndexBackfiller.java [hunk 1]
```diff
@@ -130,31 +127,28 @@ public Results backfill() {
...
   /** Writes index entries until the cap is reached. Returns the number of entries written. */
   private int writeIndexEntries(LocalDocumentsView localDocumentsView) {
-    int totalEntriesWrittenCount = 0;
+    int documentsProcessed = 0;
     Timestamp startingTimestamp = Timestamp.now();
 
-    while (totalEntriesWrittenCount < maxIndexEntriesToProcess) {
-      int entriesRemainingUnderCap = maxIndexEntriesToProcess - totalEntriesWrittenCount;
+    while (documentsProcessed < maxDocumentsToProcess) {
+      int entriesRemainingUnderCap = maxDocumentsToProcess - documentsProcessed;
```
Comment (thebrianchen) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/IndexBackfiller.java:141]: s/entriesRemaining/documentsRemaining

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/IndexBackfiller.java [hunk 2]
```diff
@@ -130,31 +127,28 @@ public Results backfill() {
...
-              /* numIndexesRemoved= */ 0);
+          int documentsProcessed = writeIndexEntries(localDocumentsView);
+          return new Results(/* hasRun= */ true, documentsProcessed);
         });
   }
 
   /** Writes index entries until the cap is reached. Returns the number of entries written. */
   private int writeIndexEntries(LocalDocumentsView localDocumentsView) {
-    int totalEntriesWrittenCount = 0;
+    int documentsProcessed = 0;
```
Comment (thebrianchen) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/IndexBackfiller.java:137]: update comment to reflect "Number of docs" written

**Review by thebrianchen — APPROVED** [review]
lgtm, the simplification to track documents instead of entries processed is awesome.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/IndexBackfiller.java (hunk 1 above)
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/IndexBackfiller.java]: Done

**Pull request merged**